src/qibo/backends/numpy.py

```diff
@@ -429,28 +432,64 @@ def set_device(self, name):
             self.set_engine("numba")
 
     def to_numpy(self, x):
+        if isinstance(x, self.np.ndarray):
+            return x
         return self.op.to_numpy(x)
 
     def cast(self, x, dtype='DTYPECPX'):
         if isinstance(dtype, str):
             dtype = self.dtypes(dtype)
         return self.op.cast(x, dtype=dtype)
 
+    def check_shape(self, shape):
+        if self.op.get_backend() == "cupy" and isinstance(shape, self.Tensor):
+            shape = shape.get()
+        return shape
+
+    def reshape(self, x, shape):
+        return super().reshape(x, self.check_shape(shape))
+
+    def eye(self, shape, dtype='DTYPECPX'):
+        return super().eye(self.check_shape(shape), dtype=dtype)
+
+    def zeros(self, shape, dtype='DTYPECPX'):
+        return super().zeros(self.check_shape(shape), dtype=dtype)
+
+    def ones(self, shape, dtype='DTYPECPX'):
```

Comment:
Just a question about this functions, if I understand correctly, the shapes should not be `self.Tensor`, but rather numpy/list/numbers right?